### DISCUSSION TITLE
Shared Memory as the additional transport [14934]

### DISCUSSION THREAD

**Discussion by sfod** [eProsima/Fast-DDS#2768] [Q&A]
The question is basically the same as [this old question](https://github.com/eProsima/Fast-DDS/issues/1665), but is about the DDS API rather than the deprecated FastRTPS API.

I've been experimenting with [hello world shared memory example](https://github.com/eProsima/Fast-DDS/tree/master/examples/C%2B%2B/DDS/HelloWorldExampleSharedMem). I've enabled both SharedMem and UDPv4 transports for a participant in a publisher.
```
    auto udp_transport = std::make_shared<UDPv4TransportDescriptor>();
    pqos.transport().user_transports.push_back(udp_transport);
```
The issue is that a subscriber participant, with only SharedMemory enabled, doesn't see the publisher. When I add the UDPv4 transport to the subscriber, they start sending data over UDP. From my understanding, when any other transport is enabled for a participant, SharedMemory transport basically becomes disabled.

Is this an intended behavior or a bug? Is it possible to use the SharedMemory transport along with other transports?

**Comment by JLBuenoLopez** [comment]
Hi @sfod,

It is possible to use Shared Memory (SHM) transport with other transports. However, there are some design limitations. If your participant has some other transport different from SHM enabled, the discovery communication is only performed using that other transport (and not SHM). Therefore, if one participant has only SHM enabled and the other has SHM and UDP, no discovery is going to happen and no communication is going to be established. Either you disable the UDP transport in the participant that has it enabled, or you enable UDP in the other participant. I have issued eProsima/Fast-DDS-docs#374 in order to clarify this.

**Comment by sfod** [comment]
Hi @JLBuenoLopez-eProsima,

Thanks for the answer and for the updated docs. Though, one thing is still unclear to me.

>  Also, when two participants on the same machine have SHM transport enabled, the user data communication between them
  is automatically performed by SHM transport only.

When I add the UDPv4 transport to the hello world example (i.e. both publisher and subscriber use SharedMemory + UDP), the communication goes over UDP.

**Reply by JLBuenoLopez** [reply]
Only the participant discovery traffic is exchanged through UDP (if you capture the traffic using some tool as wireshark you will only see DATA(p) messages but no DATAs). The user data messages will be exchanged through SHM (so wireshark will not capture them).

**Reply by sfod** [reply]
Yes, you're absolutely right. Thanks again for the help!
